Jaspr UI components

mjablecnik opened the pull request:
Hello @schultek thank you a lot for Jaspr web framework. It is very nice and I believe it can be very popular in the future :-)

Now I have an idea to create some small simple components for better, beautiful and faster development with Jaspr. 

I added another  `jaspr_ui` package into your Jaspr project with some basic HTML elements, CSS Styles and Bootstrap component. This is only simple kickoff for see how it can look and I plan to add more in the future. 
In examples you can see 4 small examples and in ROADMAP.md is simple TODO list about what is done and what I plan to do.

A lot of work is here yet and I would like to ask you if you are opened to add `jaspr_ui` into your project when I will develop and maintain this part of Jaspr. I believe it can be very useful for many people similar like Material or Cupertino widgets in Flutter.
And you will have also one more contributor in this project ;-)

Let me know what do you think about it.
Martin

schultek commented:
Wow what an effort. First of all thanks so much, I really appreciate it! ❤️

I've had a quick glance at the code but will take a closer look as soon as I got more time.

However here are a few points on that topic, some of which I think need more consideration:

1. I love the general idea of a ui package for jaspr, its would bring so much value to the package!
2. I wouldn't want such a package (and you seem to have the same view) as part of the core framework but instead as a separate package, with separate versions etc. (as opposed to how flutter does this)
3. You did both some 
    - wrapped html components (header, div, img, ...),
    - opinionated styled components (bootstrap stuff)
  
    I probably wouldn't want those inside the same package, to not confuse users. The first category could actually be part of the core framework, and the second would definitely be its separate package.
    
4. When only containing opiniated styled components, I think such a package would probably better be called 'jaspr_bootstrap' to allow for other opinionated styled (3rd-party?) component packages (e.g. 'jaspr_material', 'jaspr_bulma', etc.)

5. I'm not sure If such a package needs to be part of the core repository / project. I think it would be fine to publish this as a third party. However then I would be very open for recommending this package in the core documentation.

6. I haven't looked in detail how you did the styles. As you might know the general topic of styles and css in jaspr is still todo. Maybe you have any input for this?

If you have any ideas or opinions on these points please let me know. I'm open for further discussion. Overall this seems like a great contribution and good direction to go in.

mjablecnik commented:
Thank you very much for your fast response :-)
Here I have answers to your points:
1) Me too :)

2) My idea is to have one package for jaspr where will be core, state management and router in one jaspr package because it will be probably used in any or most projects or is here some reason to have it separated?
And packages for testing and ui have separated in other packages in this repository.

3) Yes I firstly created `jaspr_bootstrap` package but early I found that I need wrappers for basic HTML components for better readability and basic styles components are also missing. 
Hence I added it into my package and renamed it into `jaspr_ui` if you would like to split it so it is ok but I would like to keep it in same repository.

4) Yes it is ok for me. We can create in root directory of this repository `thirdparty` directory where can be projects which are only optional and are not part of jaspr core packages. There can be packages like `jaspr_bootstrap`, `jaspr_bulma`, `jaspr_material`, etc.. And also we say to other developers that we are open for next contributors into this repository which can increase popularity of this package and it is very crucial in start of any open source project.
I don't like projects where is one core repository and many other additional repositories from many developers everywhere in GitHub and you don't know which repository is really active or not. When every contributor will create their packages here so it will be very clearly which any other plugin exists how many active contributors jaspr community have and we can also control if code of third party package will not be deprecated or unmainteined in this repository. 
My big inspiration is [Godot Game Engine](https://github.com/godotengine/godot) which have also thirdparty folder in their repository and every contributor cannot find any other point to contribution into this project.
Any thirdparty package will have one maintainer of course.

5) My opinion is in point 4.

6) Yes I know that it is in TODO. In your ROADMAP I see: `Add css preprocessing and scoped styles`.
I didn't do any css preprocessing  (SCSS, SASS, Less, etc.) or scoped styles.
I only added other element wrapper for style tag where can be defined css stylesheet of the whole page. Small example is here: https://github.com/mjablecnik/jaspr/blob/bootstrap-components/packages/jaspr_ui/example/lib/pages/grid_system_page.dart#L6
For styles I am using Style class which is little similar like DomComponent for HTML tags.
And every style can be grouped into StyleGroup and added into StyleElement for set style for whole page.
Or you can also add it into any HTML element or component like inline style.
In the future it can be more involved of course.
For better understand you can study this parts of code: https://github.com/mjablecnik/jaspr/blob/bootstrap-components/packages/jaspr_ui/lib/src/core/elements/style.dart
https://github.com/mjablecnik/jaspr/tree/bootstrap-components/packages/jaspr_ui/lib/src/core/styles

schultek commented:
Hi @mjablecnik 

I think I want add the basic html and style components to the core jaspr package, but under a separate library to import, e.g. `import 'package:jaspr/ui.dart`. This way it's optional to use, but does not require a separate package dependency. I get your point and I'm fine with adding the other package as `jaspr_bootstrap` under a `thirdparty` directory in this repo.

So it would be awesome if you could make a separate pull request for the basic elements to put into the core package. 

**However:**

Unfortunately I won't be able to work on jaspr over the next 4 weeks, so no need to hurry with this. Also if you want to publish it sooner, I would propose doing it as a separate project for now where you can work on it without me, and then when I'm back we can merge it into the main project, including the basic elements.

mjablecnik commented:
Hello @schultek,
so I divided the jaspr_ui and simple ui with styles part I moved into `jaspr/ui` core library and bootstrap component part I moved into thirdparty directory.
All changes are for simplicity in this branch and after your return you can merge it into main branch.

snehmehta commented:
Hey @mjablecnik,
I was planning to something like this and I am totally aligned with your view.
If you let me, I would like to contribute at your bootstrap components branch.

mjablecnik commented:
Hello @snehmehta,
Yes of course you can fork it and create some PRs here into my bootstrap-components branch and I will to do some code review for you till @schultek will be back ;-)

Can you only let me know which tasks from my TODO lists would you like to do? :-)
https://github.com/mjablecnik/jaspr/blob/bootstrap-components/packages/jaspr/TODO.md
https://github.com/mjablecnik/jaspr/blob/bootstrap-components/thirdparty/jaspr_bootstrap/TODO.md

mjablecnik commented:
@snehmehta I created discord community server where you can join and we can communicate on the next progress there ;-)
https://discord.gg/G8pH8X4u

schultek reviewed (commented):
Can you move the example in the jaspr package to somewhere else for now? I would take some separate time to create a good helpful general example for jaspr.

If you want to experiment with what you are building, you can add this under the `experiments` folder.
Additionally you can also add a simple sample to `jasprpad`, you can look at the other samples on how to do this.

schultek commented on packages/jaspr/lib/src/ui/elements/div.dart:
```diff
@@ -0,0 +1,14 @@
+import 'package:jaspr/src/ui/elements/base.dart';
+
+class DivElement extends BaseElement {
+  const DivElement({
+    super.key,
+    super.id,
+    super.style,
+    super.classes,
+    super.attributes,
+    super.events,
+    super.child,
+    super.children,
+  }) : super(tag: 'div');
+}
```
I think these wrapper elements are then not needed when the html pr gets merged. 

We should only keep / implement the components that provide value beyond the raw html tag, like `Image`, `Table`, `ListView` etc.

mjablecnik replied:
Yes but right now I have a lot of usages of this element in `jaspr/bootstrap` library. So I only mark this as a deprecated and after merge of `jaspr/html` I will update `jaspr/bootstrap` and this element I will remove then.

schultek replied:
Makes sense

schultek commented on packages/jaspr/lib/src/ui/forms/radio_button.dart:
```diff
@@ -0,0 +1,42 @@
+import 'package:jaspr/src/ui/utils.dart';
+import 'package:jaspr/ui.dart';
+
+class RadioButton extends BaseElement {
```
I like these form input components 🤩. They need some more detailed attention, but the overall concept is great!

mjablecnik replied:
Yes I know. This is only on start and have to be made a lot of work yet.. :)

schultek commented on packages/jaspr/pubspec.yaml:
```diff
@@ -5,7 +5,7 @@ homepage: https://github.com/schultek/jaspr
 
 environment:
-  sdk: '>=2.16.0 <3.0.0'
+  sdk: '>=2.17.0 <3.0.0'
```
I updated this on the main branch so you can merge this into yours.

mjablecnik replied:
Yes I know I updated it already but now I have this problem: https://github.com/schultek/jaspr/issues/20
Can you fix it please?

schultek commented on thirdparty/jaspr_bootstrap/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
The other packages use a symlink to the topmost license, if thats ok for you I would do this here too.

mjablecnik replied:
I updated the license.

mjablecnik commented on packages/jaspr/analysis_options.yaml:
```diff
@@ -3,5 +3,5 @@ include: package:lints/recommended.yaml
 linter:
   rules:
     sort_pub_dependencies: true
-    prefer_relative_imports: true
```
I don't like relative imports I prefer write only absolute imports path.
Because sometimes you can somewhere see something like: `import '../../../../something.dart'`
For example after some refactoring or anything similar..
So better is write absolute paths everywhere I think for better readability and consistency with other imports which are not relative because are from different package for example..

schultek replied:
Absolute paths can get long too. This just comes down to personal preference. 
I would keep it as it is right now.

mjablecnik replied:
I am not talking about long of imports path but I talk about readability.

mjablecnik commented on packages/jaspr/lib/src/ui/elements/center.dart:
```diff
@@ -0,0 +1,14 @@
+    super.child,
+    super.children,
+  }) : super(tag: 'center');
+}
```
I updated it.

mjablecnik commented:
@schultek What do you think about this directory structure?:
![image](https://user-images.githubusercontent.com/2686906/179121373-5784f22b-4620-4d95-976e-54b4f1bc9a94.png)

All from UI will be in `ui` directory and in `components` will be my Flutter-like components, in `html` will be your html
methods and in `styles` will be `Style` class which can be used with your html methods or with my Flutter-like components.

So then we can have this libraries:
- `package:jaspr/components.dart`
- `package:jaspr/html.dart`
- `package:jaspr/styles.dart`

mjablecnik commented:
@schultek Otherwise.. can you please already merge this PR? I think that it is quite huge right now and if you now will also develop on this project so we should have the same directory structure and some shared files because I saw that you already copied a lot of my code into one your another branch and this is very bad practice because then can arise big conflicts and I am very concerned about it right now..

schultek commented:
I created a wip/v0.2 branch where we can continue working on.
